sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -933,51 +933,58 @@ async function getOrCreateEntryAsync(
-        const environment = entry.object.env;
+        entry.object = entry.object || { env: new Map() };
 
         if (localCapturedPropertyChains.length === 0) {
-            await serializeAllObjectPropertiesAsync(environment);
+            await serializeAllObjectPropertiesAsync(entry.object);
             return false;
         } else {
-            return await serializeSomeObjectPropertiesAsync(environment, localCapturedPropertyChains);
+            return await serializeSomeObjectPropertiesAsync(entry.object, localCapturedPropertyChains);
         }
     }
 
     // Serializes out all the properties of this object.  Used when we can't prove that
     // only a subset of properties are used on this object.
-    async function serializeAllObjectPropertiesAsync(environment: PropertyMap) {
+    async function serializeAllObjectPropertiesAsync(object: ObjectInfo) {
         // we wanted to capture everything (including the prototype chain)
         const descriptors = await getOwnPropertyDescriptors(obj);
 
         for (const descriptor of descriptors) {
-            // we're about to recurse inside this object.  In order to prever infinite
-            // loops, put a dummy entry in the environment map.  That way, if we hit
-            // this object again while recursing we won't try to generate this property.
             const keyEntry = await getOrCreateEntryAsync(getNameOrSymbol(descriptor), undefined, context, serialize, logInfo);
-            if (!environment.has(keyEntry)) {
-                environment.set(keyEntry, <any>undefined);
 
-                const propertyInfo = await createPropertyInfoAsync(descriptor);
-                const prop = await getOwnPropertyAsync(obj, descriptor);
-                const valEntry = await getOrCreateEntryAsync(
-                    prop, undefined, context, serialize, logInfo);
-
-                // Now, replace the dummy entry with the actual one we want.
-                environment.set(keyEntry, { info: propertyInfo, entry: valEntry });
+            // We're about to recurse inside this object.  In order to prevent infinite loops, put a
+            // dummy entry in the environment map.  That way, if we hit this object again while
+            // recursing we won't try to generate this property.
+            //
+            // Note: we only stop recursing if we hit exactly our sentinel key (i.e. we're self
+            // recursive).  We *do* want to recurse through the object again if we see it through
+            // non-recursive paths.  That's because we might be hitting this object through one
+            // prop-name-path, but we created it the first time through another prop-name path.
+            //
+            // By processing the object again, we will add the different members we need.
+            if (object.env.has(keyEntry) && object.env.get(keyEntry) === undefined) {
+                continue;
             }
+            object.env.set(keyEntry, <any>undefined);
```

Comment:
this is the core change to what we had before.  previously, we said "ah, we've **marker** this prop, don't go into it again".  This was done both to prevent infinite recursion, and to avoid serializing a prop we thought we'd already serialized.

Now, all we do is set the marking to prevent infinite recursion.  if we see the prop a second time otherwise, then we still re-serialize it.  This way, if we have a differnet set of sub-props to serialize that will ensure that that happens.